File: app/src/main/java/com/usher/demo/rx/RxSearchActivity.java
package com.usher.demo.rx;

import android.os.Bundle;
import androidx.annotation.Nullable;
import androidx.appcompat.app.AppCompatActivity;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;
import android.widget.EditText;

import com.jakewharton.rxbinding3.widget.RxTextView;
import com.usher.demo.R;
import com.usher.demo.utils.RxUtil;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.concurrent.Callable;
import java.util.concurrent.TimeUnit;

import io.reactivex.Observable;

public class RxSearchActivity extends AppCompatActivity {
    private final List<String> mSearchResults = new ArrayList<>();
    private final List<String> mCities = new ArrayList<>();

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        setContentView(R.layout.activity_rx_search);

        initData();
        initView();
    }

    private void initData() {
        mCities.addAll(Arrays.asList("北京市", "上海市", "天津市", "重庆市", "沈阳市", "台北市"));
    }

    private void initView() {
        RxSearchAdapter adapter = new RxSearchAdapter(this, mSearchResults);

        RecyclerView recyclerView = findViewById(R.id.recyclerview);
        recyclerView.setLayoutManager(new LinearLayoutManager(this, RecyclerView.VERTICAL, false));
        recyclerView.setAdapter(adapter);


        EditText input = findViewById(R.id.edittext);
        RxTextView.textChanges(input)
                .debounce(500, TimeUnit.MILLISECONDS)
                .map(v -> v.toString().trim())
                .distinctUntilChanged()
                .switchMap(v -> Observable.fromIterable(mCities)
                        .filter(city -> city.contains(v))
                        .collect((Callable<ArrayList<String>>) ArrayList::new, ArrayList::add)
                        .toObservable()
                )
                .compose(RxUtil.getSchedulerComposer())
                .as(RxUtil.autoDispose(this))
                .subscribe(v -> {
                    mSearchResults.clear();
                    mSearchResults.addAll(v);
                    adapter.notifyDataSetChanged();
                });
    }
}

File: app/src/main/java/com/usher/demo/rx/RxSearchActivity.kt
package com.usher.demo.rx

import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.chad.library.adapter.base.BaseViewHolder
import com.jakewharton.rxbinding4.widget.textChanges
import com.twigcodes.ui.adapter.RxBaseQuickAdapter
import com.twigcodes.ui.util.RxUtil
import com.usher.demo.R
import com.usher.demo.base.BaseActivity
import io.reactivex.rxjava3.core.Observable
import kotlinx.android.synthetic.main.activity_rx_search.*
import kotlinx.android.synthetic.main.item_search.view.*
import java.util.concurrent.TimeUnit

class RxSearchActivity : BaseActivity(R.layout.activity_rx_search) {
    private val mCities = listOf("北京市", "上海市", "天津市", "重庆市", "沈阳市", "台北市")
    private val mAdapter by lazy { SearchAdapter(listOf()) }

    private val getCities: (keyword: String) -> Observable<List<String>> = { keyword ->
        Observable.create { emitter ->
            val cities = mCities.filter { city -> city.contains(keyword) }
            emitter.onNext(cities)
        }
    }

    override fun initView() {
        recyclerview.layoutManager = LinearLayoutManager(this, RecyclerView.VERTICAL, false)
        recyclerview.adapter = mAdapter

        edittext.textChanges()
                .debounce(500, TimeUnit.MILLISECONDS)
                .map { it.trim().toString() }
                .distinctUntilChanged()
                .switchMap { getCities(it) }
                .compose(RxUtil.getSchedulerComposer())
                .to(RxUtil.autoDispose(this))
                .subscribe {
                    mAdapter.setNewData(it)
                }
    }

    private class SearchAdapter(data: List<String>) : RxBaseQuickAdapter<String, BaseViewHolder>(R.layout.item_search, data) {

        override fun convert(holder: BaseViewHolder, item: String) {
            holder.itemView.name_textview.text = item
        }
    }
}